### ISSUE TITLE
provide clearer advice about USVString vs. DOMString

### ISSUE THREAD

**Issue by dbaron** [heycam/webidl#84]
In https://github.com/w3ctag/spec-reviews/issues/87#issuecomment-171535440 (see also the following comment) we had a brief discussion about USVString vs. DOMString.  This seems like a somewhat tricky (in that it's easy to get wrong) API design issue, and the current wording in WebIDL isn't particularly clear.

http://heycam.github.io/webidl/#idl-USVString currently says:
> Specifications SHOULD only use USVString for APIs that perform text processing and need a string of Unicode scalar values to operate on. Most APIs that use strings should instead be using DOMString, which does not make any interpretations of the code units in the string. When in doubt, use DOMString.

This is a bit unclear in a few ways (not clear what "text processing" means; does encoding conversion count?), and doesn't seem to match the other advice being given, e.g., in https://github.com/w3ctag/spec-reviews/issues/87#issuecomment-171575546

This seems like an area where giving clear and consistent advice is important.

**Comment by annevk** [comment]
The plan is to rephrase this in terms of https://infra.spec.whatwg.org/#strings and use JavaScript (for DOMString) and scalar value string (for USVString) from there. Is the difference sufficiently clear there?

**Comment by SimonSapin** [comment]
@annevk I think you’re response is about *what* each of the two types are, while @dbaron was asking about *why* and in what situations is one recommended over the other. I agree that it is unclear where “APIs that perform text processing” applies.

**Comment by annevk** [comment]
No, it specifically calls out UTF-8 encode as a place where USVString applies. I haven't really countered another place, though the source does mention Rust subsystems as something where this might make sense, but since Rust isn't used on a large scale that's not super relevant yet.

**Comment by domenic** [comment]
So concretely, I think we should point out that USVString should be used for URLs or for strings that are to be sent "over the wire". I'm not sure what other scenarios there might be.

**Comment by SimonSapin** [comment]
For what it’s worth, Firefox’s ongoing style system rewrite uses UTF-8 strings internally. This effectively makes CSSOM strings be `USVString`. CSSWG resolved to allow both, so https://github.com/w3c/csswg-drafts/pull/1266 introduced a `CSSOMString` typedef.

**Comment by domenic** [comment]
Right... I don't think we want to propogate such implementation-specific quirks throughout the platform via Web IDL advice.

**Comment by SimonSapin** [comment]
Right, `CSSOMString` probably does not have its place in this WebIDL spec note. This was only a response to "not sure what other scenarios there might be".

**Comment by jyasskin** [comment]
Is https://w3ctag.github.io/design-principles/#idl-string-types a sufficient fix for this?

**Comment by cynthia** [comment]
I believe so. We might have just forgot to close this.

**Comment by jan-ivar** [comment]
I'd like clearer advice on whether to use USVString for read-only attributes like [url](https://w3c.github.io/webrtc-pc/#dom-rtcicecandidate-url) in https://github.com/w3c/webrtc-pc/issues/3029.

Personally, I find such WebIDL annotations on output helpful (as a shortcut to reading algorithms) even though I know they don't translate to processing steps the way they do on input. E.g. much like I'd expect a method to never return a dictionary without its `required` members filled in, I'd expect a read-only `USVString` attribute to be free of surrogates.

> Is https://w3ctag.github.io/design-principles/#idl-string-types a sufficient fix for this?

It seems preoccupied with input, except _"...strings whose most common algorithms operate on [scalar values](https://infra.spec.whatwg.org/#scalar-value)"_ which suggests yes (use USVString on the read-only [url](https://w3c.github.io/webrtc-pc/#dom-rtcicecandidate-url) attribute).

But https://webidl.spec.whatwg.org/#idl-USVString warns no: _"... only use [USVString](https://webidl.spec.whatwg.org/#idl-USVString) for APIs that perform text processing and need a string of [scalar values](https://infra.spec.whatwg.org/#scalar-value) to operate on. Most APIs that use strings should instead be using [DOMString](https://webidl.spec.whatwg.org/#idl-DOMString), which does not make any interpretations of the [code units](https://infra.spec.whatwg.org/#code-unit) in the string. When in doubt, use [DOMString](https://webidl.spec.whatwg.org/#idl-DOMString)."_ — this seems to match @domenic's https://github.com/w3ctag/design-principles/issues/93#issuecomment-379816152: _"Only use USVString when it saves you a step of standards text, by rolling that step into the Web IDL type conversion."_

Can we clarify the guideline and align the warning with it?

**Comment by domenic** [comment]
I would caution against generally using `USVString` on outputs as a sort of comment, similar to how we don't encourage using `required` in that way. However, saying that all URL-valued strings should be `USVString`, including in outputs, seems like a good idea for consistency and for reducing spec maintainer overhead.
